Add numeric StringComparator

jon-wei opened the pull request:
This PR adds a pure numeric StringComparator (supports decimals, negative numbers), used by the BoundFilter and GroupBy/TopN/Search queries.

The comparator will attempt to parse and compare Strings as BigDecimals.

Unparseable strings are treated as nulls, and nulls precede numbers. When comparing two strings that cannot be parsed as numbers, the comparator will order them lexicographically.

The BoundDimFilter interface has also been changed, deprecating "alphanumeric" in favor of an "ordering" String field.

The PR also unifies the existing lexicographic and alphanumeric TopNMetricSpecs (i.e., the ones that sort on dimension values) into a DimensionTopNMetricSpec, which can accept any of the 4 string comparators currently supported.

A similar unifying change is made for the SearchSortSpec (implemented in NewSearchSortSpec).

Fixes https://github.com/druid-io/druid/issues/2989

===========

The following benchmarks show the performance difference between using this numeric comparator and the existing alphanumeric comparator on a long column (__time).

The `basic` data set spans an interval from (0,1000000) milliseconds.

`timeFilter` runs a timeseries query on the entire interval, using a filtered aggregator long sum with a BoundFilter for the __time interval (200000, 300000). The BoundFilter has an optimization that compares using long primitives if possible, when using the numeric comparator.

`timeFilterAlphanumeric` is the same as the previous query, but using the existing alphanumeric comparator instead.

`timeFilterByInterval` is a timeseries query on the interval (200000, 300000) with a long sum. This query is meant to provide a baseline comparison (interval based filtering vs. boundfilter on __time column). The interval based filtering is faster since the IncrementalIndex/QueryableIndex have optimizations that allow exclusion of rows using the interval (the boundfilter applied to the whole interval still has to read all of the rows)

```
TimeseriesBenchmark.queryMultiQueryableIndex                       1            750000              basic.timeFilter  avgt   50    19433.259 ±    66.505  us/op
TimeseriesBenchmark.queryMultiQueryableIndex                       1            750000  basic.timeFilterAlphanumeric  avgt   50   108962.170 ±  1877.408  us/op
TimeseriesBenchmark.queryMultiQueryableIndex                       1            750000    basic.timeFilterByInterval  avgt   50     7578.105 ±    20.135  us/op

TimeseriesBenchmark.querySingleIncrementalIndex                    1            750000              basic.timeFilter  avgt   50    69655.796 ±  2193.242  us/op
TimeseriesBenchmark.querySingleIncrementalIndex                    1            750000  basic.timeFilterAlphanumeric  avgt   50   144758.532 ±  3640.986  us/op
TimeseriesBenchmark.querySingleIncrementalIndex                    1            750000    basic.timeFilterByInterval  avgt   50    14898.808 ±   467.090  us/op

TimeseriesBenchmark.querySingleQueryableIndex                      1            750000              basic.timeFilter  avgt   50    19668.441 ±    60.053  us/op
TimeseriesBenchmark.querySingleQueryableIndex                      1            750000  basic.timeFilterAlphanumeric  avgt   50   101008.642 ±  1595.471  us/op
TimeseriesBenchmark.querySingleQueryableIndex                      1            750000    basic.timeFilterByInterval  avgt   50     7490.932 ±    15.981  us/op
```


The following benchmark shows performance of the numeric comparator vs. alphanumeric in a TopN query:

```
Benchmark                                  (numSegments)  (rowsPerSegment)        (schemaAndQuery)  (threshold)  Mode  Cnt       Score       Error  Units
TopNBenchmark.queryMultiQueryableIndex                 1            750000       basic.numericSort           10  avgt   25  782488.365 ± 10153.742  us/op
TopNBenchmark.queryMultiQueryableIndex                 1            750000  basic.alphanumericSort           10  avgt   25  713468.221 ± 12023.384  us/op
TopNBenchmark.querySingleIncrementalIndex              1            750000       basic.numericSort           10  avgt   25  588548.252 ±  7134.410  us/op
TopNBenchmark.querySingleIncrementalIndex              1            750000  basic.alphanumericSort           10  avgt   25  534799.964 ±  7137.141  us/op
TopNBenchmark.querySingleQueryableIndex                1            750000       basic.numericSort           10  avgt   25  391563.687 ±  6699.905  us/op
TopNBenchmark.querySingleQueryableIndex                1            750000  basic.alphanumericSort           10  avgt   25  351552.575 ±  7352.647  us/op
```

jon-wei commented:
rebasing this on top of #3180

jon-wei commented:
rebased

acslk commented on docs/content/querying/sorting-orders.md:
```diff
@@ -0,0 +1,31 @@
+When comparing two unparseable values (e.g., "hello" and "world"), this ordering will sort by comparing the unparsed strings lexicographically.
+
+## Strlen
+Sorts values by the their string lengths.
```
I think it would be helpful to also define the ordering for strings with same length when using Strlen. It seems that it uses the natural ordering right now, which should be lexicographical. Is there any actual difference between Ordering.natural and our Lexicographic Ordering implementation?

jon-wei replied:
I'll add a note that Strlen falls back to "natural" ordering on ties.

Ordering.natural will just call String's compareTo method, the lexicographic implementation in druid converts the strings to UTF8 byte arrays and compares on those. Not sure whether these comparators order Strings differently practically.

acslk commented on processing/src/main/java/io/druid/query/filter/BoundDimFilter.java:
```diff
@@ -135,12 +178,15 @@ public ExtractionFn getExtractionFn()
+        + orderingBytes.length
     );
     boundCacheBuffer.put(DimFilterUtils.BOUND_CACHE_ID)
                     .put(boundType)
```
Might as well remove the AlphaNumericByte for cache key since it's not used for equal or hashCode.

acslk replied:
GetDimensionRangeSet of BoundDimFilter (for #2982) only supports Lexicographical ordering right now, so the if(... || alphaNumeric) { return null;} should be changed to if(... || !ordering.equals(lexicographic))

jon-wei replied:
thanks, missed that

gianm commented:
@jon-wei could you please post results from the new benchmarks?

jon-wei commented:
@gianm added benchmarks to main comment

acslk commented:
👍

fjy commented on processing/src/main/java/io/druid/query/filter/BoundDimFilter.java:
```diff
@@ -50,8 +57,9 @@ public BoundDimFilter(
-      @JsonProperty("extractionFn") ExtractionFn extractionFn
+      @Deprecated @JsonProperty("alphaNumeric") Boolean alphaNumeric,
+      @JsonProperty("extractionFn") ExtractionFn extractionFn,
+      @JsonProperty("ordering") String ordering
```
can we make this into an enum that has a method fromString(String ordering)?

jon-wei replied:
@fjy Replaced this field with a new StringComparatorType enum

gianm commented on docs/content/querying/filters.md:
```diff
@@ -174,26 +174,33 @@ The IN filter supports the use of extraction functions, see [Filtering with Extr
+|upper|String|The upper bound for the filter|no|
+|lowerStrict|Boolean|Perform strict comparison on the lower bound ("<" instead of "<=")|no, default: false|
+|upperStrict|Boolean|Perform strict comparison on the upper bound (">" instead of ">=")|no, default: false|
+|ordering|String|Specifies the sorting order to use when comparing values against the bound.<br/>Can be one of the following values: "lexicographic", "alphanumeric", "numeric", "strlen".<br/>See [Sorting Orders](./sorting-orders.html) for more details.|no, default: "lexicographic"|
```
Something is messing up the formatting here, maybe the embedded html (check out "View", the table doesn't render). Even if it's not the culprit, I don't think the `<br/>` are really needed, we could allow the browser's builtin word wrap do its thing.

jon-wei replied:
@gianm Fixed this, the problem was the `|--------|-----------|---------|` line that needed another column.

Also removed the `<br/>` tags.

gianm commented on docs/content/querying/sorting-orders.md:
```diff
@@ -0,0 +1,31 @@
+These sorting orders are used by the [TopNMetricSpec](./topnmetricspec.html), [SearchQuery](./searchquery.html), GroupByQuery's [LimitSpec](./limitspec.html), and [BoundFilter](./filters.html#bound-filter).
+
+## Lexicographic
+Sorts values in case-sensitive lexicographic order.
```
We can be more specific here. This sorts the UTF-8 representation of a string in lexicographic order, on a byte by byte basis.

jon-wei replied:
@gianm added details on UTF-8 and byte-by-byte lexicographic comparison

gianm commented on docs/content/querying/sorting-orders.md:
```diff
@@ -0,0 +1,31 @@
+## Alphanumeric
+Suitable for strings with both numeric and non-numeric content, e.g.:
+
+"file12 sorts after file2"
```
This might look better with fewer line breaks.

jon-wei replied:
@gianm reduced line breaks

gianm commented on processing/src/main/java/io/druid/query/topn/NumericDimensionTopNMetricSpec.java:
```diff
@@ -0,0 +1,77 @@
+import java.util.Comparator;
+import java.util.List;
+
+public class NumericDimensionTopNMetricSpec extends LexicographicTopNMetricSpec
```
This should be able to take any ordering, no need to limit it to numeric.

gianm replied:
Would also probably make sense to flip the inheritance, i.e. the lexicographic spec is subclass of this which enables additional optimizations. Or, collapse them into one class, and have the optimization chosen based on the comparator.

gianm commented:
@jon-wei looks like there are compilation issues in TopNQueryRunnerTest.

gianm commented on processing/src/main/java/io/druid/query/filter/BoundDimFilter.java:
```diff
@@ -40,8 +46,10 @@
   private final boolean lowerStrict;
   private final boolean upperStrict;
-  private final boolean alphaNumeric;
+  private final Boolean alphaNumeric;
```
alphaNumeric doesn't need to be a class field, it only needs to be understood in the constructor and translated to an ordering as appropriate (allowing us to migrate parameters this way is why http://druid.io/docs/latest/operations/rolling-updates.html recommends updating historical/MMs before brokers).

jon-wei replied:
@gianm removed unneeded field

gianm commented on processing/src/main/java/io/druid/query/filter/BoundDimFilter.java:
```diff
@@ -173,7 +220,7 @@ public Filter toFilter()
   public RangeSet<String> getDimensionRangeSet(String dimension)
   {
-    if (!Objects.equals(getDimension(), dimension) || getExtractionFn() != null || alphaNumeric) {
+    if (!Objects.equals(getDimension(), dimension) || getExtractionFn() != null || !ordering.equals(StringComparatorType.LEXICOGRAPHIC)) {
```
hmm, this would probably be easier to read as a NOT of ANDs rather than an OR of NOTs.

jon-wei replied:
@gianm changed this to NOT of ANDs

gianm commented on processing/src/main/java/io/druid/query/ordering/StringComparatorType.java:
```diff
@@ -0,0 +1,49 @@
+  {
+    return new byte[]{(byte) this.ordinal()};
+  }
+}
```
Newline at end of file

jon-wei replied:
@gianm file has been deleted

gianm commented on processing/src/main/java/io/druid/query/ordering/StringComparators.java:
```diff
@@ -342,16 +355,86 @@ public String toString()
+    }
+  }
+
+  public static StringComparator makeComparator(StringComparatorType type)
```
I think we can actually get rid of `StringComparatorType` and hook this directly up to Jackson. Try annotating this with `@JsonCreator` and either putting it inside `StringComparator` (which would have to be an abstract class).

See QueryGranularity for an example of a class that works this way. It can accept both strings (like `"minute"` and objects (like `{"type":"period", ...}`)

gianm replied:
That should simplify things a bit; there's no need to have this switch, or isOrderingValid, or the StringComparatorType enum, or for callers to call a conversion method.

jon-wei replied:
@gianm using jackson for StringComparator now

fjy commented on processing/src/main/java/io/druid/query/filter/BoundDimFilter.java:
```diff
@@ -60,8 +69,31 @@ public BoundDimFilter(
-    this.alphaNumeric = (alphaNumeric == null) ? false : alphaNumeric;
+    this.alphaNumeric = alphaNumeric;
+
+    if (ordering == null) {
```
can we add a note about what is for backwards compatibility?

jon-wei replied:
@fjy added note on alphanumeric usage/deprecation

jon-wei commented:
still making jackson changes

fjy commented on processing/src/main/java/io/druid/query/ordering/StringComparator.java:
```diff
@@ -0,0 +1,45 @@
+public abstract class StringComparator implements Comparator<String>
+{
+  @JsonCreator
+  public static StringComparator fromString(String type)
```
is this method still needed? we should just be able to use jackson

jon-wei replied:
@fjy it's needed to create the StringComparator from a simple string, e.g. `alphanumeric`, vs `{"type": "alphanumeric"}`

fjy replied:
ah okay

fjy commented on processing/src/main/java/io/druid/query/groupby/orderby/OrderByColumnSpec.java:
```diff
@@ -202,7 +202,7 @@ private static StringComparator determinDimensionComparator(Object dimensionOrde
 
     String dimensionOrderString = dimensionOrderObj.toString().toLowerCase();
-    return StringComparators.makeComparator(dimensionOrderString);
+    return StringComparator.fromString(dimensionOrderString);
```
can we use jackson here?

jon-wei replied:
@fjy I think that'll require some more changes to the way OrderByColumnSpec is constructed, can we handle this in another PR?

jon-wei replied:
@fjy Refactored OrderByColumnSpec to use jackson more with latest comit

jon-wei commented:
I had reverted the PR to before I added the "StringComparatorType" enum, which removed some changes I made to the TopNMetricSpec, let me bring those files back in

jon-wei commented:
added new unified TopNMetricSpec and SearchSortSpec

jon-wei commented:
added docs for DimensionTopNMetricSpec, updated main comment

gianm commented on processing/src/main/java/io/druid/query/search/search/NewSearchSortSpec.java:
```diff
@@ -0,0 +1,101 @@
+
+import java.util.Comparator;
+
+public class NewSearchSortSpec
```
It looks like SearchSortSpec is not used for anything any longer, why not replace it with this rather than having two different types?

jon-wei replied:
@gianm I think I was being unnecessarily conservative about deleting the old classes, got rid of them now

gianm commented on processing/src/main/java/io/druid/query/search/search/NewSearchSortSpec.java:
```diff
@@ -0,0 +1,101 @@
+  {
+    return ordering.hashCode();
+  }
+}
```
Newline at end of file

jon-wei replied:
@gianm added newline

gianm commented on processing/src/main/java/io/druid/query/topn/DimensionTopNMetricSpec.java:
```diff
@@ -0,0 +1,182 @@
+public class DimensionTopNMetricSpec implements TopNMetricSpec
+{
+  private static final StringComparator DEFAULT_ORDERING = StringComparators.LEXICOGRAPHIC;
+  public static final byte STRING_SEPARATOR = (byte) 0xFF;
```
Could be private

jon-wei replied:
@gianm made private

gianm commented on processing/src/main/java/io/druid/query/topn/DimensionTopNMetricSpec.java:
```diff
@@ -0,0 +1,182 @@
+  public byte[] getCacheKey()
+  {
+    byte[] previousStopBytes = previousStop == null ? new byte[]{} : StringUtils.toUtf8(previousStop);
+    byte[] orderingBytes = StringUtils.toUtf8(ordering.toString());
```
I think this line means that StringComparator needs a getCacheKey method.

jon-wei replied:
@gianm added getCacheKey() method to StringComparator

gianm commented on processing/src/main/java/io/druid/query/topn/DimensionTopNMetricSpec.java:
```diff
@@ -0,0 +1,182 @@
+    result = 31 * result + getOrdering().hashCode();
+    return result;
+  }
+}
```
Newline

jon-wei replied:
@gianm added newline

gianm commented on processing/src/main/java/io/druid/query/ordering/StringComparators.java:
```diff
@@ -342,16 +329,61 @@ public String toString()
-    } else {
-      throw new IAE("Unknown string comparator[%s]", type);
+    @Override
+    public int compare(String o1, String o2)
```
@jon-wei Do you have benchmarks on how this performs relative to alphaNumeric for dimensions that only contain integers (in which they should order the same way)? If not could you please add one to the topN benchmarks?

I'm wondering if we need to think about optimizing this for numerics that fit in primitives or not. If the performance is roughly similar I think we don't need to worry about it.

jon-wei replied:
@gianm I've added two new queries to the TopNBenchmark that use numeric/alphanumeric sort.

Benchmark results have been posted to the main comment.

gianm commented on docs/content/querying/topnmetricspec.md:
```diff
@@ -28,39 +28,26 @@ The metric field can also be given as a JSON object. The grammar for dimension v
 |metric|the actual metric field in which results will be sorted by|yes|
 
-## Lexicographic TopNMetricSpec
+## Dimension TopNMetricSpec
```
I think it'd be good to include a small note that lexicographic and alphanumeric specs are still supported for backwards compatibility, but are deprecated. Not a full section - just a note to reassure people that they won't stop working after an upgrade.

jon-wei replied:
@gianm Added note here about deprecated lexicographic and alphanumeric specs

vogievetsky replied:
I have to say that the name `TopNMetricSpec` is kind of ridiculous at this point.  It was always a little silly but now it is a step worse. it should really be `TopNSortSpec` or similar.

so instead of writing `"metric": { "type": "dimension", ... }` you could write `"sort": { "type": "dimension", ... }`

Maybe this should not be changed as part of this PR. Maybe this will never be changed.

gianm replied:
@vogievetsky, probably not this PR (it's already long) although in general I would be down with adding `ordering` as a synonym for `metric` in topN and adjusting the docs to call them "Orderings" instead of "TopNMetricSpecs".

gianm commented on processing/src/main/java/io/druid/query/search/search/SearchSortSpec.java:
```diff
@@ -19,22 +19,83 @@
+  {
+    return ordering.hashCode();
+  }
+}
```
Newline at end of file

jon-wei replied:
@gianm added newline

gianm commented on processing/src/main/java/io/druid/query/search/search/SearchSortSpec.java:
```diff
@@ -19,22 +19,83 @@
+
+  @JsonCreator
+  public SearchSortSpec(
+      @JsonProperty("ordering") StringComparator ordering
```
I think this needs to be called `"type"` not `"ordering"` in order for this to be backwards compatible.

jon-wei replied:
@gianm renamed this to `type`

navis commented on processing/src/main/java/io/druid/query/ordering/StringComparators.java:
```diff
@@ -340,18 +345,81 @@ public String toString()
+      Long long1 = o1 == null ? null : Longs.tryParse(o1);
+      Long long2 = o2 == null ? null : Longs.tryParse(o2);
+
+      final BigDecimal bd1 = long1 == null ? convertStringToBigDecimal(o1) : new BigDecimal(long1);
```
we can avoid unnecessary NPE for null value (o1, o2)

gianm replied:
yes, good point, perhaps `convertStringToBigDecimal` should have a short-circuit "if input is null then return null"

jon-wei replied:
sounds good, added a null check to `convertStringToBigDecimal`

jon-wei replied:
also added some null checks to the NumericComparator before the attempted numeric conversions

fjy commented:
@jon-wei some merge conflicts

jon-wei commented:
@fjy rebased

fjy commented:
👍

gianm commented on processing/src/main/java/io/druid/query/groupby/orderby/OrderByColumnSpec.java:
```diff
@@ -42,50 +41,74 @@
-      throw new ISE("Cannot build an OrderByColumnSpec from a %s", obj.getClass());
-    }
+  @JsonCreator
+  public static OrderByColumnSpec fromString(String dimension)
```
Are there serde tests for this class? Having two `@JsonCreator` on the same class seems like the kind of thing that might work or might not work.

jon-wei replied:
@gianm DefaultLimitSpecTest had a serde test, I've added more tests there, the two @JsonCreator functions work

gianm commented on processing/src/main/java/io/druid/query/ordering/StringComparator.java:
```diff
@@ -17,29 +17,31 @@
  * under the License.
  */
 
-package io.druid.query.search;
```
I wonder what made git think that StrlenSearchSortSpecTest.java was renamed to StringComparator.java

gianm commented on processing/src/test/java/io/druid/query/ordering/StringComparatorsTest.java:
```diff
@@ -120,27 +120,61 @@ public void testStrlenComparator()
+    Assert.assertTrue(StringComparators.NUMERIC.compare("-500000000.14124", "CAN'T TOUCH THIS") > 0);
+    Assert.assertTrue(StringComparators.NUMERIC.compare("CAN'T PARSE THIS", "-500000000.14124") < 0);
+
+    Assert.assertTrue(StringComparators.NUMERIC.compare("CAN'T PARSE THIS", "CAN'T TOUCH THIS") < 0);
```
🔨 🕙

gianm commented:
👍
